sidecar/tlv.go

```diff
@@ -0,0 +1,377 @@
+
+		if err := tlv.DUint64(r, &tmp, buf, 8); err != nil {
+			return err
+		}
+		v.PushAmt = btcutil.Amount(tmp)
+
+		if err := DPubKeyN(r, &v.SignPubKey, buf); err != nil {
+			return err
+		}
+
+		return DSigN(r, &v.SigOfferDigest, buf)
+	}
+	return tlv.NewTypeForEncodingErr(val, "*Offer")
+}
+
+// ERecipient is an Encoder for the Recipient type. An error is returned if val
+// is not a **Recipient.
+func ERecipient(w io.Writer, val interface{}, buf *[8]byte) error {
+	if vPtr, ok := val.(**Recipient); ok {
+		v := *vPtr
+
+		if err := EPubKeyN(w, &v.NodePubKey, buf); err != nil {
+			return err
+		}
+
+		return EPubKeyN(w, &v.MultiSigPubKey, buf)
+	}
+	return tlv.NewTypeForEncodingErr(val, "*Recipient")
+}
+
+// DRecipient is a Decoder for the Recipient type. An error is returned if val
+// is not a **Recipient.
+func DRecipient(r io.Reader, val interface{}, buf *[8]byte, l uint64) error {
+	if vPtr, ok := val.(**Recipient); ok && l == recipientLength {
+		v := &Recipient{}
+		*vPtr = v
+
+		if err := DPubKeyN(r, &v.NodePubKey, buf); err != nil {
+			return err
+		}
+
+		return DPubKeyN(r, &v.MultiSigPubKey, buf)
+	}
+	return tlv.NewTypeForEncodingErr(val, "*Recipient")
+}
+
+// EOrder is an Encoder for the Order type. An error is returned if val is not
+// a **Order.
+func EOrder(w io.Writer, val interface{}, buf *[8]byte) error {
+	if vPtr, ok := val.(**Order); ok {
+		v := *vPtr
+
+		if err := ENonce(w, &v.BidNonce, buf); err != nil {
+			return err
+		}
+
+		return ESigN(w, &v.SigOrderDigest, buf)
+	}
+	return tlv.NewTypeForEncodingErr(val, "*Order")
+}
+
+// DOrder is a Decoder for the Order type. An error is returned if val is not
+// a **Order.
+func DOrder(r io.Reader, val interface{}, buf *[8]byte, l uint64) error {
+	if vPtr, ok := val.(**Order); ok && l == orderLength {
+		v := &Order{}
+		*vPtr = v
+
+		if err := DNonce(r, &v.BidNonce, buf); err != nil {
+			return err
+		}
+
+		return DSigN(r, &v.SigOrderDigest, buf)
+	}
+	return tlv.NewTypeForEncodingErr(val, "*Order")
+}
+
+// EExecution is an Encoder for the Execution type. An error is returned if val
+// is not a **Execution.
+func EExecution(w io.Writer, val interface{}, buf *[8]byte) error {
+	if vPtr, ok := val.(**Execution); ok {
+		v := *vPtr
+
+		return tlv.EBytes32(w, &v.PendingChannelID, buf)
+	}
+	return tlv.NewTypeForEncodingErr(val, "*Execution")
+}
+
+// DExecution is a Decoder for the Execution type. An error is returned if val
+// is not a **Execution.
+func DExecution(r io.Reader, val interface{}, buf *[8]byte, l uint64) error {
+	if vPtr, ok := val.(**Execution); ok && l == executionLength {
+		v := &Execution{}
+		*vPtr = v
+
+		return tlv.DBytes32(r, &v.PendingChannelID, buf, 32)
+	}
+	return tlv.NewTypeForEncodingErr(val, "*Execution")
+}
+
+// EPubKeyN is a nil safe Encoder for the btcec.PublicKey type. An error is
+// returned if val is not a **btcec.PublicKey.
+func EPubKeyN(w io.Writer, val interface{}, buf *[8]byte) error {
```

Comment:
kinda feels like a lot of boiler plate here. Shouldn't most of the logic we need (writing bytes to a TLV stream) already be defined somewhere it can be reused?

Reply:
There are pre-defined functions for many things available, yes. But there wasn't one for conditionally serializing a pubkey depending on whether it was nil or not. But I think we should instead use individual TLV records everywhere then the nilness problem would go away.